### Assessing Data
In this step,the data is to be understood more deeply,Before implementing methods to clean it ,you will definetly need to have a better idea about what the data is about

### Types of Unclean Data
+ Dirty Data(data with quality issues): -> known as low quality data.Low quality data has content issues
+ Messy Data(data with tidlness issues) -> Messy data,also known as untidy data.Untidy data has structueral issues



### 1. Dirty data
When the data has following kinds of issues
+ 1. Duplicated data
+ 2. Missing Data
+ 3. Corrupt data
+ 4. inaccurate data

### 2. Messy data
tidy data has the following propertis

+ 1. Each variables forms a column
+ 2. Each observation forms a row
+ 3. Each observation unit forms a table

### Types of Assessment
+ 1. Manual-Looking through the data manually in goole sheets
+ 2. Programmatic-By using pandas functions such as info(),describe() or sample()

### steps in assessment
+ 1. Discover
+ 2. Document

### Data Quality Dimensions


In [4]:
import pandas as pd

In [5]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse = pd.read_csv('adverse_reactions.csv')

In [6]:
patients.shape

(503, 14)

In [7]:
patients.head(1)

patient_id assigned_sex given_name  surname               address  \
0           1       female        Zoe  Wellish  576 Brown Bear Drive   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   

                                contact  birthdate  weight  height   bmi  
0  951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6

In [8]:
treatment.shape

(280, 7)

In [9]:
treatment.head(1)

given_name   surname    auralin novodra  hba1c_start  hba1c_end  \
0   veronika  jindrová  41u - 48u       -         7.63        7.2   

   hba1c_change  
0           NaN

In [10]:
adverse.shape

(34, 3)

In [11]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

### Quality
[Completeness, Validity, Accuracy, Consistency ]
### Patients -> 
+ zip code col is stored in decimal format [v] , 
+ ceratin zip codes of 4 digit [v],
+ for the state column sometimes full name are used and other time short form used [ic]
+ Missing values for contact details of 12 patients [com]
+ incorrect data type for assigned sex and birth date [?]
+ Errorneous data for weights and heights [ac]
+ Inconsistent way of storing phone numbers in contact col [ic]
+ Duplicates and not useful data in datset [ic]
### Treatments -> 
+ hba1c_change has missing value[comp]
+ Novodra and Auralin cols has u attached to numbers [val]
+ only 280 row are there insted of 375 [ic]
+ name and sirname col is in lower case[ic]
+ incorrect data  type for auralin and novadra cols[va]
+ hba1c_change 9 instead of 4[ac]
+ use of dashes insted of nan in aurolin and novodira cols[ic]

### Adverse Reaction ->
+ name and sirname col is in lower case[ic]


In [12]:
## programmatic assessment
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [13]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [14]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [15]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [16]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [17]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [18]:
patients.sample()


patient_id assigned_sex given_name    surname             address    city  \
70          71         male  Alexander  Mathiesen  3130 Jessie Street  Athens   

   state  zip_code        country  \
70    OH   45701.0  United States   

                                            contact  birthdate  weight  \
70  AlexanderEMathiesen@superrito.com1 740 629 5172  12/2/1955   195.6   

    height   bmi  
70      68  29.7

In [19]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

### Tidiness
### Patients
+ phone and email together in one col i.e contact
###  Treatments
+ 3 variables stored in auralin and novodra cols
### Adverse Reaction
+ no point of using this table

### steps in Assessment
### There are 2 steps involved in Assessment
+ 1. Discover
+ 2. Document


### Data Quality Dimensions
+ 1. Completeness issues
+ 2. Validity Issuess
+ 3. Consistency Issues
+ 4. Accuracy Issues

### Cleaning Data
You should follow the order provided below
+ Completeness issues
+ Tidness issues
+ Remaining Data Quality issues like validity,accuracy and consistency

### Steps involved in Data Cleaning
+ Define
+ Code
+ Test

Always make sure to create a copy of your pandas dataframe before you start the cleaning process



In [21]:
# Creating copies
patients_copy = patients.copy()
treatment_copy = treatment.copy()
adverse_copy = adverse.copy()

In [22]:
patients[patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

### a. Completeness Issuess
+ Missing values
+ Ohnly 280 rows are there
+ hba1c_change col has missing value

### Define

In [23]:
# Substract haba1c_start fromhaba1c_end to get hba1c_change

### Code

In [24]:
treatment_copy['hba1c_change'] = treatment_copy['hba1c_start'] - treatment_copy['hba1c_end']

### Test

In [31]:
treatment_copy['hba1c_change'].isnull().sum()

0

In [32]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  280 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [33]:
treatment_copy.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    280.000000
mean      7.985929    7.589286      0.396643
std       0.568638    0.569672      0.060168
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.350000
50%       7.800000    7.420000      0.390000
75%       7.970000    7.570000      0.440000
max       9.950000    9.580000      0.540000

### b. Tidiness Issues
+ Phone+email in contact col patients
+ Novodra and Aurolin col treatment
+ Merge the adverse table to treatment table


### Define

In [ ]:
# Novadra and aurolin using the melt function

### Code

In [36]:
treatment_copy = pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
        value_vars=['auralin', 'novodra'],
        var_name='treatment', value_name='dose')

In [38]:
treatment_copy = treatment_copy[treatment_copy['dose'] !='-']

In [40]:
treatment_copy[['start', 'end']] = treatment_copy['dose'].str.split('-', expand=True)

In [43]:
treatment_copy.drop(columns={'dose'},inplace=True)

In [44]:
treatment_copy.head(1)

given_name   surname  hba1c_start  hba1c_end  hba1c_change treatment start  \
0   veronika  jindrová         7.63        7.2          0.43   auralin  41u    

    end  
0   48u

### Test